packages/openapi3/src/openapi-spec-mappings.ts

```diff
@@ -0,0 +1,79 @@
+import { Namespace, Program } from "@typespec/compiler";
+import { TypeEmitter } from "@typespec/compiler/emitter-framework";
+import { MetadataInfo } from "@typespec/http";
+import { getExternalDocs, resolveInfo } from "@typespec/openapi";
+import { OpenAPI3EmitterOptions, OpenAPIVersion } from "./lib.js";
+import { ResolvedOpenAPI3EmitterOptions } from "./openapi.js";
+import { createWrappedOpenAPI3SchemaEmitterClass } from "./schema-emitter-3-0.js";
+import { createWrappedOpenAPI31SchemaEmitterClass } from "./schema-emitter-3-1.js";
+import { SupportedOpenAPIDocuments } from "./types.js";
+import { VisibilityUsageTracker } from "./visibility-usage.js";
+import { XmlModule } from "./xml-module.js";
+
+export interface OpenApiSpecSpecificProps {
+  createRootDoc: (
+    program: Program,
+    serviceType: Namespace,
+    serviceVersion?: string,
+  ) => SupportedOpenAPIDocuments;
+
+  /**
+   * Creates a constructor for the schema emitter class for the given OpenAPI version.
+   * This constructor is meant to be used when creating an AssetEmitter.
+   */
+  createSchemaEmitterCtor: (
+    metadataInfo: MetadataInfo,
+    visibilityUsage: VisibilityUsageTracker,
+    options: ResolvedOpenAPI3EmitterOptions,
+    xmlModule: XmlModule | undefined,
+  ) => typeof TypeEmitter<Record<string, any>, OpenAPI3EmitterOptions>;
+}
+
+export function getOpenApiSpecProps(specVersion: OpenAPIVersion): OpenApiSpecSpecificProps {
+  switch (specVersion) {
+    case "3.0.0":
+      return {
+        createRootDoc(program, serviceType, serviceVersion) {
+          return createRoot(program, serviceType, specVersion, serviceVersion);
+        },
+        createSchemaEmitterCtor: createWrappedOpenAPI3SchemaEmitterClass,
```

Comment:
I'm not sure I fully understand what you're suggesting - is it to have `createSchemaEmitter` return the `AssetEmitter` directly instead of the constructor that's passed to the current `createAssetEmitter` function? (Not sure what `sh` is).

I can do that - so instead of:
```ts
schemaEmitter = createAssetEmitter(
      program,
      createSchemaEmitterCtor(metadataInfo, visibilityUsage, options, xmlModule),
      context,
    );
```
you'd have
```ts
schemaEmitter = createSchemaEmitter(program, context, metadataInfo, visibilityUsage, options, xmlModule);
```

Might need to actually use named props instead given the number of positional args but is that what you meant here?

Reply:
The `AssetEmitter` needs to be passed a constructor though - you can't give it an instance of a class.

Reply:
```ts
//...
case "3.0.0": 
  default: {  
    return () => {
      const createOpenAPI3Schema = () => ({root: ...});  
      const oas3Emiter = new OpenAPI3SchemaEmitter(sh);  
      const emitter = new OpenAPI31SchemaEmitter(oas3Emiter);  
    
    };  
  }  
//...
```

Reply:
//...  
case "3.0.0": 
  default: {  
    return () => {  
      const createOpenAPI3Schema = () => ({root: ...});  
      const oas3Emiter = new OpenAPI3SchemaEmitter(sh);  
      const emitter = new OpenAPI31SchemaEmitter(oas3Emiter);  
      return emitter
    };  
  }  
//...

Reply:
I see, there's a limit in emitter framework of `createAssetEmitter` function, it needs constructor.
I feel like passing an interface to create type emitter is a common pattern? So that we don't need to extend OpenAPI31SchemaEmitter in order to pass the constructor. I feel like customers also face the same difficulty to create a type emitter with more parameters.
In the below example, the interface is `typeEmitterCreator: TypeEmitterCreator<T, TOptions>`

```ts
// packages/openapi3/src/schema-emitter-3-1.ts
export class OpenAPI31SchemaEmitter extends OpenAPI3SchemaEmitterBase<OpenAPISchema3_1> { 
  constructor(
    assetEmitter: AssetEmitter<OpenAPI3Schema, OpenAPI3EmitterOptions>, 
    metadataInfo: MetadataInfo,
    visibilityUsage: VisibilityUsageTracker,
    options: ResolvedOpenAPI3EmitterOptions,
    xmlModule: XmlModule | undefined
  ) {}
}

class OpenAPI3TypeEmitterCreator implements TypeEmitterCreator<OpenAPI3Schema, OpenAPI3EmitterOptions> {
  constructor(
    assetEmitter: AssetEmitter<OpenAPI3Schema, OpenAPI3EmitterOptions>, 
    metadataInfo: MetadataInfo,
    visibilityUsage: VisibilityUsageTracker,
    options: ResolvedOpenAPI3EmitterOptions,
    xmlModule: XmlModule | undefined
  ) {}
  create: (assetEmitter) => new OpenAPI3Emitter(this.assetEmitter, this.metadataInfo, this.visibilityUsage, this.options, this.xmlModule);
};

// packages/openapi3/src/openapi-spec-mappings.ts
export function getOpenApiSpecProps(specVersion: OpenAPIVersion): OpenApiSpecSpecificProps {
  switch (specVersion) {
    case "3.1.0": {
      return new openAPI3TypeEmitterCreator(assetEmitter, metadataInfo, visibilityUsage, options, xmlModule);
    }
    case "3.0.0": 
    default: {
      // ...
    }
  }
}

// packages/compiler/src/emitter-framework/asset-emitter.ts
interface TypeEmitterCreator<T, TOptions extends object> { 
  create: (assetEmitter: AssetEmitter<T, TOptions>) => TypeEmitter<T, TOptions>
}

export function createAssetEmitter<T, TOptions extends object>(
  program: Program,
  //TypeEmitterClass: typeof TypeEmitter<T, TOptions>,
  typeEmitterCreator: TypeEmitterCreator<T, TOptions>,
  emitContext: EmitContext<TOptions>,
): AssetEmitter<T, TOptions> {
  // ...
  const typeEmitter_ = typeEmitterCreator.create(assetEmitter);
  // ...

```

Reply:
Oh sure - I don't disagree that the pattern I've seen is to pass a factory function instead of an actual constructor, but `createAssetEmitter` is part of the entity framework and not really something that should be changed as part of this PR.